### PR TITLE
python3Packages.einops: fixed tests

### LINKED ISSUE #152049: python3Packages.einops tests are broken
`python3Package.einops` is effectively broken both in nixos-21.11 and master: tests take too long to finish.

Steps to reproduce:

* Run `nix-shell -p 'python3.withPackages(ps: with ps; [ einops ])'`
* Observe apparently stuck tests

Build log:
```
building '/nix/store/fwm8kamh7hxd88zjq27hzcvkimc9d1cv-python3.9-einops-0.3.2.drv'...
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Using setuptoolsShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing setuptools-check-hook
unpacking sources
unpacking source archive /nix/store/ly1jyc21vjvkpn52spljywdpxlanc6gm-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/tests/test_parsing.py
patching sources
configuring
no configure script, doing nothing
building
Executing setuptoolsBuildPhase
running bdist_wheel
running build
running build_py
creating build
creating build/lib
creating build/lib/einops
copying einops/parsing.py -> build/lib/einops
copying einops/einops.py -> build/lib/einops
copying einops/_backends.py -> build/lib/einops
copying einops/__init__.py -> build/lib/einops
creating build/lib/einops/layers
copying einops/layers/torch.py -> build/lib/einops/layers
copying einops/layers/tensorflow.py -> build/lib/einops/layers
copying einops/layers/keras.py -> build/lib/einops/layers
copying einops/layers/gluon.py -> build/lib/einops/layers
copying einops/layers/chainer.py -> build/lib/einops/layers
copying einops/layers/_weighted_einsum.py -> build/lib/einops/layers
copying einops/layers/__init__.py -> build/lib/einops/layers
installing to build/bdist.linux-x86_64/wheel
running install
running install_lib
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/wheel
creating build/bdist.linux-x86_64/wheel/einops
creating build/bdist.linux-x86_64/wheel/einops/layers
copying build/lib/einops/layers/__init__.py -> build/bdist.linux-x86_64/wheel/einops/layers
copying build/lib/einops/layers/_weighted_einsum.py -> build/bdist.linux-x86_64/wheel/einops/layers
copying build/lib/einops/layers/chainer.py -> build/bdist.linux-x86_64/wheel/einops/layers
copying build/lib/einops/layers/gluon.py -> build/bdist.linux-x86_64/wheel/einops/layers
copying build/lib/einops/layers/keras.py -> build/bdist.linux-x86_64/wheel/einops/layers
copying build/lib/einops/layers/tensorflow.py -> build/bdist.linux-x86_64/wheel/einops/layers
copying build/lib/einops/layers/torch.py -> build/bdist.linux-x86_64/wheel/einops/layers
copying build/lib/einops/__init__.py -> build/bdist.linux-x86_64/wheel/einops
copying build/lib/einops/_backends.py -> build/bdist.linux-x86_64/wheel/einops
copying build/lib/einops/einops.py -> build/bdist.linux-x86_64/wheel/einops
copying build/lib/einops/parsing.py -> build/bdist.linux-x86_64/wheel/einops
running install_egg_info
running egg_info
creating einops.egg-info
writing einops.egg-info/PKG-INFO
writing dependency_links to einops.egg-info/dependency_links.txt
writing top-level names to einops.egg-info/top_level.txt
writing manifest file 'einops.egg-info/SOURCES.txt'
reading manifest file 'einops.egg-info/SOURCES.txt'
adding license file 'LICENSE'
writing manifest file 'einops.egg-info/SOURCES.txt'
Copying einops.egg-info to build/bdist.linux-x86_64/wheel/einops-0.3.2-py3.9.egg-info
running install_scripts
adding license file "LICENSE" (matched pattern "LICEN[CS]E*")
creating build/bdist.linux-x86_64/wheel/einops-0.3.2.dist-info/WHEEL
creating 'dist/einops-0.3.2-py3-none-any.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
adding 'einops/__init__.py'
adding 'einops/_backends.py'
adding 'einops/einops.py'
adding 'einops/parsing.py'
adding 'einops/layers/__init__.py'
adding 'einops/layers/_weighted_einsum.py'
adding 'einops/layers/chainer.py'
adding 'einops/layers/gluon.py'
adding 'einops/layers/keras.py'
adding 'einops/layers/tensorflow.py'
adding 'einops/layers/torch.py'
adding 'einops-0.3.2.dist-info/LICENSE'
adding 'einops-0.3.2.dist-info/METADATA'
adding 'einops-0.3.2.dist-info/WHEEL'
adding 'einops-0.3.2.dist-info/top_level.txt'
adding 'einops-0.3.2.dist-info/RECORD'
removing build/bdist.linux-x86_64/wheel
Finished executing setuptoolsBuildPhase
installing
Executing pipInstallPhase
/build/source/dist /build/source
Processing ./einops-0.3.2-py3-none-any.whl
Installing collected packages: einops
Successfully installed einops-0.3.2
/build/source
Finished executing pipInstallPhase
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/igdnm1b9jf7w86by0v8d07mqlsvdchyk-python3.9-einops-0.3.2
strip is /nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0/bin/strip
stripping (with command strip and flags -S) in /nix/store/igdnm1b9jf7w86by0v8d07mqlsvdchyk-python3.9-einops-0.3.2/lib
patching script interpreter paths in /nix/store/igdnm1b9jf7w86by0v8d07mqlsvdchyk-python3.9-einops-0.3.2
checking for references to /build/ in /nix/store/igdnm1b9jf7w86by0v8d07mqlsvdchyk-python3.9-einops-0.3.2...
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
running install tests
/build/source/tests/__init__.py:60: UserWarning: backend could not be initialized for tests: <class 'einops._backends.JaxBackend'>
  warnings.warn('backend could not be initialized for tests: {}'.format(backend_type))
:param symbolic: symbolic or imperative frameworks? ... /build/source/tests/__init__.py:60: UserWarning: backend could not be initialized for tests: <class 'einops._backends.JaxBackend'>
  warnings.warn('backend could not be initialized for tests: {}'.format(backend_type))
ok
:param symbolic: symbolic or imperative frameworks? ... ok
tests.test_einsum.test_all ... /build/source/einops/layers/_weighted_einsum.py:49: FutureWarning: WeightedEinsum is experimental feature. API can change in unpredictable and enjoyable ways
  warnings.warn('WeightedEinsum is experimental feature. API can change in unpredictable and enjoyable ways',
/build/source/einops/layers/_weighted_einsum.py:86: UserWarning: WeightedEinsum: weight has no dimensions (means multiplication by a number)
  warnings.warn('WeightedEinsum: weight has no dimensions (means multiplication by a number)')
ok
tests.test_examples.test_rearrange_examples ... 2021-12-24 21:51:44.048870: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
ok
:param symbolic: symbolic or imperative frameworks? ... ok
tests.test_examples.test_pytorch_yolo_fragment ... ok
:param symbolic: symbolic or imperative frameworks? ... ok
tests.test_layers.test_rearrange_imperative ... ok
tests.test_layers.test_rearrange_symbolic ... ok
tests.test_layers.test_reduce_imperative ... /nix/store/sr711v3q725xxiybwnlfx6167s1zhqy4-python3.9-numpy-1.21.2/lib/python3.9/site-packages/numpy/core/_methods.py:52: RuntimeWarning: overflow encountered in reduce
  return umr_prod(a, axis, dtype, out, keepdims, initial, where)
ok
tests.test_layers.test_reduce_symbolic ... ok
```

### PR DESCRIPTION
**Pull request by yl3dy**
###### Motivation for this change

Fix occasionally stuck tests, see #152049.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by yl3dy** [comment]
Result of `nixpkgs-review pr 152291` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python38Packages.einops</li>
    <li>python39Packages.einops</li>
  </ul>
</details>

File: pkgs/development/python-modules/einops/default.nix [hunk 1]
```diff
@@ -43,8 +43,15 @@ buildPythonPackage rec {
...
-    export HOME=$TMPDIR
+    # Leftover files in a $TMPDIR can screw tests up, so creating a separate dir
+    export home_backup=$HOME
+    export HOME=$(mktemp -d)
+
     nosetests -v -w tests
+
+    rm -rf $HOME
+    export HOME=$home_backup
+    unset home_backup
```
Comment (risicle) [inline pkgs/development/python-modules/einops/default.nix:54]: I'm not sure you really have to bother tidying up after yourself or setting `HOME` back to its previous setting. Many packages set `HOME` in the `checkPhase` and I'm not sure I've ever seen any bother setting it back.

Comment (risicle) [reply pkgs/development/python-modules/einops/default.nix:54]: Also, completely replaced `checkPhase`s could do with `runHook preCheck` and `runHook postCheck` and the beginning and end respectively.

Comment (legendofmiracles) [reply pkgs/development/python-modules/einops/default.nix:54]: I've been wondering about that, why is that needed? Does python add something to those phases, or is it for easier overriding?

Comment (risicle) [reply pkgs/development/python-modules/einops/default.nix:54]: The hooks? AFAIK It's purely for the convenience of people who may need to override elements of the package and e.g. set their own extra `preCheck`, expecting it to work rather than being ignored.

Comment (yl3dy) [reply pkgs/development/python-modules/einops/default.nix]: > Many packages set HOME in the checkPhase and I'm not sure I've ever seen any bother setting it back.

It was more of a precaution, so removed it.

> Also, completely replaced checkPhases could do with runHook preCheck and runHook postCheck and the beginning and end respectively.

Added, thanks a lot for a suggestion!

**Comment by veprbl** [comment]
@GrahamcOfBorg build python3Packages.einops

**Comment by veprbl** [comment]
The tests seem to hang on test_torch_layer on ofborg and in my local tests. It seems like all tests are passing within 3m on the upstream github actions CI.

**Comment by yl3dy** [comment]
> The tests seem to hang on test_torch_layer on ofborg and in my local tests. It seems like all tests are passing within 3m on the upstream github actions CI.

What system do you use? I tried building via `nixpkgs-review` tens of times with the proposed fixes on NixOS 21.11 x86_64 (don't have a darwin system unfortunately), and the whole test suite ran for less than 1m. I'm quite at loss here: tried my best to isolate tests, but don't have a lot of experience with PyTorch to determine what could possibly go wrong here. On the other hand, I'd rather not disable this test.

**Comment by veprbl** [comment]
@yl3dy I was testing on non-nixos x86_64-linux running on an i7-3770K with sandboxing enabled.

**Comment by samuela** [comment]
I just tried running nixpkgs-review on this PR and it's still hung up right after `test_reduce_symbolic ... ok`. I'm running an AWS EC2 m5.24xlarge instance. 
```
❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.87, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.11.334578.75f3534c98a, nix-ld"`
 - channels(skainswo): `"home-manager, nixpkgs-unstable-22.05pre343295.adf7f03d3bf"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```
I believe it should be reproducible running on any of the m5 instances (at least).

@yl3dy What kind of CPU are you running OOC? Anything else unusual about your setup?

**Comment by yl3dy** [comment]
I'm running Core i7-12700K at NixOS 21.11 (stable channel), no dGPU. Running on CPU might be the most unusual, compared to non-Nix testing.
